File: problem-1/sol.py
import sys
import os
import heapq

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))) 
from aoc_input import get_aoc_input


def calculate_total_distance():
    input_data = get_aoc_input(day=1)

    left_list = []
    right_list = []

    # Prepare the left and right list
    for line in input_data.splitlines():
        left, right = map(int, line.split())
        left_list.append(left)
        right_list.append(right)

    heapq.heapify(left_list)
    heapq.heapify(right_list)

    res = 0

    while left_list and right_list:
        left_min = heapq.heappop(left_list)
        right_min = heapq.heappop(right_list)
        res += abs(left_min - right_min)

    print(f"Total distance: {res}")

calculate_total_distance()

File: problem-1/sol.go
package main

import (
	"aoc2024/aoc_input"
	"fmt"
	"log"
	"math"
	"sort"
	"strings"
)

/*
	Alternatively, one could also use a heap here
*/
func part1(inputData string) {
	lines := strings.Split(strings.TrimSpace(inputData), "\n")

	var leftList, rightList []int

	for _, line := range lines {
		var left, right int
		_, err := fmt.Sscanf(line, "%d %d", &left, &right)
		if err != nil {
			log.Fatalf("Failed to parse line: %s", line)
		}
		leftList = append(leftList, left)
		rightList = append(rightList, right)
	}

	sort.Ints(leftList)
	sort.Ints(rightList)

	totalDistance := 0
	for len(leftList) > 0 && len(rightList) > 0 {
		leftMin := leftList[0]
		rightMin := rightList[0]

		leftList = leftList[1:]
		rightList = rightList[1:]

		totalDistance += int(math.Abs(float64(leftMin - rightMin)))
	}

	fmt.Printf("Total distance: %d\n", totalDistance)
}

/*
	Bruteforcing is O(n*m) where n is the length of leftList and m is the length of rightList.
	Alternatively, we could sort the rightList and perform a binary search, in which case the complexity would be O(n*log(m))
*/

func part2(inputData string) {
	var leftList, rightList []int

	for _, line := range strings.Split(strings.TrimSpace(inputData), "\n") {
		var left, right int
		_, err := fmt.Sscanf(line, "%d %d", &left, &right)
		if err != nil {
			log.Fatalf("Failed to parse line: %s", line)
		}
		leftList = append(leftList, left)
		rightList = append(rightList, right)
	}
	rightCount := make(map[int]int)
	for _, right := range rightList {
		rightCount[right]++
	}
	similarityScore := 0
	for _, left := range leftList {
		similarityScore += left * rightCount[left]
	}
	fmt.Printf("Similarity score: %d\n", similarityScore)

}

func main() {
	inputData, err := aoc_input.GetAocInput(1, 2024)
	if err != nil {
		log.Fatalf("Failed to get AoC input: %v", err)
	}
	part1(inputData)
	part2(inputData)
}
